driver-core/src/main/com/mongodb/internal/connection/IamAuthenticator.java

```diff
@@ -0,0 +1,275 @@
+            bytes = new byte[buffer.size()];
+            int curPos = 0;
+            for (ByteBuf cur : buffer.getByteBuffers()) {
+                System.arraycopy(cur.array(), cur.position(), bytes, curPos, cur.limit());
+                curPos += cur.position();
+            }
+            return bytes;
+        }
+
+        @NonNull
+        String getUserName() {
+            String userName = credential.getUserName();
+            if (userName == null) {
+                userName = BsonDocument
+                        .parse(getHttpResponse())
+                        .getString("AccessKeyId")
+                        .getValue();
+            }
+            return userName;
+        }
+
+        @NonNull
+        private String getPassword() {
+            char[] password = credential.getPassword();
+            if (password == null) {
+                password = BsonDocument
+                        .parse(getHttpResponse())
+                        .getString("SecretAccessKey")
+                        .getValue()
+                        .toCharArray();
+            }
+            return new String(password);
+        }
+
+        @Nullable
+        private String getSessionToken() {
+            String token = credential.getMechanismProperty("AWS_SESSION_TOKEN", null);
+            if (credential.getPassword() == null || credential.getUserName() == null) {
+                if (token == null) {
+                    token = BsonDocument
+                            .parse(getHttpResponse())
+                            .getString("Token")
+                            .getValue();
+                } else {
+                    throw new IllegalArgumentException("The connection string contains auth properties and no username and password");
+                }
+            }
+            return token;
+        }
+
+
+        @NonNull
+        private String getHttpResponse() {
+            if (httpResponse == null) {
+                final String ec2 = "http://169.254.169.254/latest/meta-data/iam/security-credentials/";
+                final String ecs = "http://169.254.170.2";
+
+                String path = System.getenv("AWS_CONTAINER_CREDENTIALS_RELATIVE_URI");
+                String uri = ecs + path;
+                if (path == null) {
+                    uri = ec2 + /* role name */ getHttpContents(ec2);
+                }
+
+                httpResponse = getHttpContents(uri);
+            }
+            return httpResponse;
+        }
+
+        @NonNull
+        private static String getHttpContents(final String endpoint) {
+            StringBuffer content = new StringBuffer();
+            HttpURLConnection conn = null;
+            try {
+                conn = (HttpURLConnection) new URL(endpoint).openConnection();
+                conn.setRequestMethod("GET");
+
+                BufferedReader in = new BufferedReader(new InputStreamReader(conn.getInputStream(), Charset.forName("UTF-8")));
```

Comment:
Let's add a call to conn.getResponseCode and throw if not equal to HttpURLConnection.HTTP_OK.

Reply:
Done